op-challenger/game/keccak/matrix/matrix.go

```diff
@@ -17,10 +18,61 @@ type StateMatrix struct {
 }
 
 var (
-	ErrInvalidMaxLen = errors.New("invalid max length to absorb")
-	uint256Size      = 32
+	ErrInvalidMaxLen            = errors.New("invalid max length to absorb")
+	ErrIncorrectCommitmentCount = errors.New("incorrect number of commitments for input length")
+	uint256Size                 = 32
 )
 
+// VerifyPreimage checks the claimed state commitments are correct for the provided preimage data.
+// Returns nil if the preimage is valid, otherwise returns a [types.Challenge] proving the fault.
+func VerifyPreimage(data io.Reader, commitments []common.Hash) (*types.Challenge, error) {
+	s := NewStateMatrix()
+	m := s.PackState()
+	var prestate types.Leaf
+	for i := 0; ; i++ {
+		unpaddedLeaf, err := s.absorbNextLeafInput(data)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("failed to verify inputs: %w", err)
+		}
+		isEOF := errors.Is(err, io.EOF)
+		validCommitment := s.StateCommitment()
+		if i >= len(commitments) {
+			// There should have been more commitments.
+			// The contracts should prevent this so it can't be challenged, return an error
+			return nil, ErrIncorrectCommitmentCount
+		}
+		claimedCommitment := commitments[i]
+
+		var paddedLeaf [types.BlockSize]byte
+		copy(paddedLeaf[:], unpaddedLeaf)
```

Comment:
Good idea, though I'm still not quite sure how we're going to extract the padded version - I'd really like to be able to grab it sensibly from the existing keccak code rather than implementing it again ourselves, but I'm not sure it will be possible.  If we can just grab it, then `absorbNextLeafInput` will probably have to do that and return both padded and unpadded versions.  If we have to reimplement it, the `Pad()` function sounds ideal.

I tried out using `type unpaddedLeafData []byte` and having `absorbNextLeafInput` return that, but the compiler automatically converts it to `[]byte` everywhere it's used so it doesn't make any practical difference if we don't add a `Pad()` method. We could use a struct wrapper instead of just a type alias, but that feels like adding overhead for little value.

So if it's ok with you, I'll leave it as just `byte[]` for now but keep this in mind for when I'm working out how to handle the padding.

Reply:
Yes, completely. And we've been consistently using `[]byte` anyways in the interim, so I think it's perfectly inline for now.

I created a ticket to be sure we track this :) Please add any context there or correct me where I may be overlooking something! https://github.com/ethereum-optimism/client-pod/issues/499